File: p17471.cpp
#include<bits/stdc++.h>
using namespace std;
#define MAX 14
int N, Ecnt, E, p1, p2, ans = 1e9, cnt1, cnt2;
vector<int> people(MAX, 0), visited(MAX, 0);
vector<vector<int>> TREE(MAX, vector<int>());
vector<int> group(MAX, 0);

void dfs(int node, int flag){
    visited[node] = 1;
    if(flag == 1) cnt1++, p1 += people[node];
    else cnt2++, p2 += people[node];
    for(int i = 0; i < TREE[node].size(); i++){
        if(group[TREE[node][i]] != flag) continue;
        if(visited[TREE[node][i]] == 0) dfs(TREE[node][i], flag);
    }
}

int main(){
    cin >> N;
    for(int i = 1; i <= N; i++){
        cin >> people[i];
    }
    for(int i = 1; i <= N; i++){
        cin >> Ecnt;
        for(int j = 0; j < Ecnt; j++){
            cin >> E;
            TREE[i].push_back(E);
            TREE[E].push_back(i);
        }
    }
    for(int i = 1; i < (1 << N) - 1; i++){
        p1 = p2 = cnt1 = cnt2 = 0;
        int s1, s2;
        fill(visited.begin(), visited.end(), 0);
        fill(group.begin(), group.end(), 1);
        for(int idx = 0; idx < N; idx++){
            if((1 << idx) & i)  s1 = idx + 1;
            else s2 = idx + 1, group[idx + 1] = 2;
        }
        dfs(s1, 1);
        dfs(s2, 2);
        if(cnt1 + cnt2 == N) {
            ans = min(ans, abs(p1 - p2));
        }
    }
    if(ans == 1e9) ans = -1;
    cout << ans << "\n";
}

File: p17471.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;

public class p17471 {
    public static int N, res = Integer.MAX_VALUE;
    public static int p1, p2, cnt1, cnt2;
    public static int[] people;
    public static boolean[] group;
    public static boolean[] visited;

    public static ArrayList<ArrayList<Integer>> map = new ArrayList<>();
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        N = Integer.parseInt(br.readLine());
        people = new int[N + 1];
        String[] input = br.readLine().split(" ");

        for(int i = 0; i < N + 1; i++){
            map.add(new ArrayList<Integer>());
        }

        for(int i = 0; i < N; i++){
            people[i + 1] = Integer.parseInt(input[i]);
        }

        for(int i = 0; i < N; i++){
            input = br.readLine().split(" ");
            int M = Integer.parseInt(input[0]);
            
            for(int j = 1; j <= M; j++){
                map.get(i + 1).add(Integer.parseInt(input[j]));
            }
        }

        for(int i = 1; i < (1 << N) - 1; i++){
            p1 = p2 = cnt1 = cnt2 = 0;
            int s1 = 0, s2 = 0;
            visited = new boolean[N + 1];
            group = new boolean[N + 1];

            for(int idx = 0; idx < N; idx++){
                
                if(((1 << idx) & i) != 0) {
                    group[idx + 1] = true;
                    s1 = idx + 1;
                }
                else{
                    s2 = idx + 1;
                }
            }

            dfs(s1, true);
            dfs(s2, false);

            if(cnt1 + cnt2 != N) continue;

            res = Math.min(res, Math.abs(p1 - p2));
        }
        
        if(res == Integer.MAX_VALUE) res = -1;
        System.out.println(res);
    }

    public static void dfs(int here, boolean groupFlag){
        visited[here] = true;

        if(groupFlag){
            cnt1++;
            p1 += people[here];
        }
        else{
            cnt2++;
            p2 += people[here];
        }

        for(int i = 0; i < map.get(here).size(); i++){
            int there = map.get(here).get(i);
            if(groupFlag == group[there] && !visited[there]) dfs(there, groupFlag);
        }
    }
}
